### PR TITLE
[Tree] Fix flat tree query

### LINKED ISSUE #357: [Tree] Flat tree query issue
The `getFlatTreeQB()` method is using only LIKE operator to compare paths and can return wrong results.

Lets asume we use IDs for path generation. When getting flat tree for entity with id === 1, the query is something like `WHERE materialized_path LIKE '/1%'`. That means if i have element assigned to parent with id 10, 11, 12 and so on, it will be returned either.

### PR DESCRIPTION
**Pull request by Einenlum**
Fixes #357

### CODE REVIEW

File: src/ORM/Tree/Tree.php [hunk 1]
```diff
@@ -50,19 +51,34 @@ public function getRootNodes($rootAlias = 't')
...
-    public function getTree($path = '', $rootAlias = 't')
-    {
-        $results = $this->getFlatTree($path, $rootAlias);
+    public function getTree(
+        $path = '',
+        $rootAlias = 't',
+        $materializedPathSeparator = null
+    ) {
+        $materializedPathSeparator = $materializedPathSeparator
+            ?? self::getMaterializedPathSeparator()
```
Comment (NiR-) [inline src/ORM/Tree/Tree.php]: * This repo still requires php >= 5.4.
* If the class using this trait has `getMaterializedPathSeparator()`, for what use case is `$materializedPathSeparator` useful?

File: src/ORM/Tree/Tree.php [hunk 2]
```diff
@@ -50,19 +51,34 @@ public function getRootNodes($rootAlias = 't')
...
         return $this->buildTree($results);
     }
 
-    public function getTreeExceptNodeAndItsChildrenQB(NodeInterface $entity, $rootAlias = 't')
+    public function getTreeExceptNodeAndItsChildrenQB(NodeInterface $entity, $rootAlias = 't', $materializedPathSeparator = null)
     {
-        return $this->getFlatTreeQB('', $rootAlias)
-            ->andWhere($rootAlias.'.materializedPath NOT LIKE :except_path')
+        $materializedPathSeparator = $materializedPathSeparator
+            ?? self::getMaterializedPathSeparator()
```
Comment (NiR-) [inline src/ORM/Tree/Tree.php]: Same here

File: src/ORM/Tree/Tree.php [hunk 3]
```diff
@@ -50,19 +51,34 @@ public function getRootNodes($rootAlias = 't')
...
+        ;
+
+        $qb = $this->getFlatTreeQB('', $rootAlias, $materializedPathSeparator);
+
+        return $qb
+            ->andWhere($rootAlias.'.materializedPath NOT LIKE :except_exact_path')
+            ->andWhere($rootAlias.'.materializedPath NOT LIKE :except_parent_path')
```
Comment (NiR-) [inline src/ORM/Tree/Tree.php]: You add two `WHERE NOT LIKE` clauses here but you use the QB coming from `getFlatTreeQB` which adds a `WHERE (LIKE ... OR LIKE ...)` clause. It's a bit strange imho.

File: src/ORM/Tree/Tree.php [hunk 4]
```diff
@@ -94,12 +110,27 @@ public function buildTree($results)
...
-        $qb = $this->createQueryBuilder($rootAlias)
-            ->andWhere($rootAlias.'.materializedPath LIKE :path')
+        $materializedPathSeparator = $materializedPathSeparator
+            ?? self::getMaterializedPathSeparator()
+        ;
+
+        $qb = $this->createQueryBuilder($rootAlias);
+
+        $qb
+            ->andWhere(
```
Comment (NiR-) [inline src/ORM/Tree/Tree.php]: Maybe we can even avoid this clause when `$path` is an empty.

File: src/ORM/Tree/Tree.php [hunk 5]
```diff
@@ -94,12 +110,27 @@ public function buildTree($results)
      *
      * @return QueryBuilder
      */
-    public function getFlatTreeQB($path = '', $rootAlias = 't')
+    public function getFlatTreeQB($path = '', $rootAlias = 't', $materializedPathSeparator = null)
     {
-        $qb = $this->createQueryBuilder($rootAlias)
-            ->andWhere($rootAlias.'.materializedPath LIKE :path')
+        $materializedPathSeparator = $materializedPathSeparator
```
Comment (NiR-) [inline src/ORM/Tree/Tree.php]: Same as previous comments about `$materializedPathSeperator`.

File: src/ORM/Tree/Tree.php [hunk 6]
```diff
@@ -130,12 +161,26 @@ protected function addFlatTreeConditions(QueryBuilder $qb)
      *
      * @return array the flat resultset
      */
-    public function getFlatTree($path, $rootAlias = 't')
+    public function getFlatTree($path, $rootAlias = 't', $materializedPathSeparator = null)
     {
+        $materializedPathSeparator = $materializedPathSeparator
```
Comment (NiR-) [inline src/ORM/Tree/Tree.php]: Same again

**Comment by TomasVotruba** [comment]
If this is ~~rebased~~ sent (there were so many changes) on current `master` + covered with test, I'd be happy to merge :+1:

**Pull request closed without merging**